tools/fft.c

```diff
@@ -0,0 +1,381 @@
+ * limitations under the License.
+ */
+#include <inttypes.h>
+#include <fftw3.h>
+#include <errno.h>
+#include <stdlib.h>
+#include <string.h>
+#include <math.h>
+
+#include "../src/msg.h"
+
+#include "msg.h"
+#include "sig.h"
+#include "util.h"
+#include "fifo.h"
+
+uint32_t DEFAULT_WINDOW_LENGTH = 1000; // measured in milliseconds
+// Decided to have a fixed window interval of half the sample window.
+uint16_t DEFAULT_WINDOW_COUNT = 3; // number of windows to average together
+
+struct sample_window {
+	double *in_buffer;
+	fftw_complex *out_buffer;
+	uint32_t buffer_capacity;
+	uint32_t out_capacity;
+	uint32_t sample_count;
+	fftw_plan plan;
+	bool transformed;
+};
+
+struct channel_data {
+	uint16_t channel;
+	struct pfifo *windows;
+	uint32_t welch_window_count;
+	uint32_t windows_capacity;
+	uint32_t window_length;
+	uint32_t new_window_interval; // Samples before opening a window
+	uint32_t sample_index; // Samples since last opening a window
+	double *welch_output;
+};
+
+struct sample_window *create_sample_window(const struct channel_data *channel)
+{
+	struct sample_window *window = calloc(1, sizeof(*window));
+	if (window == NULL) {
+		return NULL;
+	}
+	window->buffer_capacity = channel->window_length;
+	window->out_capacity = channel->window_length / 2 + 1;
+	window->in_buffer = malloc(window->buffer_capacity *
+			sizeof(*window->in_buffer));
+	if (window->in_buffer == NULL) {
+		goto cleanup_window;
+	}
+	window->out_buffer = fftw_malloc((window->out_capacity) *
+			sizeof(*window->out_buffer));
+	if (window->out_buffer == NULL) {
+		goto cleanup_in;
+	}
+	window->plan = fftw_plan_dft_r2c_1d(channel->window_length,
+			window->in_buffer, window->out_buffer, FFTW_MEASURE);
+	if (window->plan == NULL) {
+		goto cleanup_out;
+	}
+	return window;
+
+cleanup_out:
+		fftw_free(window->out_buffer);
+cleanup_in:
+		free(window->in_buffer);
+cleanup_window:
+		free(window);
+		return NULL;
+}
+
+void destroy_sample_window(struct sample_window *window)
+{
+	fftw_destroy_plan(window->plan);
+	fftw_free(window->out_buffer);
+	free(window->in_buffer);
+	free(window);
+}
+
+static inline bool window_is_full(struct sample_window *window)
+{
+	return window->sample_count == window->buffer_capacity;
+}
+
+unsigned add_sample_to_window(struct sample_window *window, double sample)
+{
+	if (window->sample_count < window->buffer_capacity) {
+		window->in_buffer[window->sample_count] = sample;
+		window->sample_count++;
+	}
+	if (window_is_full(window)) {
+		if (!window->transformed) {
+			fftw_execute(window->plan);
+		}
+		return 1;
+	}
+	return 0;
+}
+
+static inline void mod_welch_average(struct channel_data *channel,
+		unsigned window_value, struct sample_window *window, int mod)
+{
+	for (uint32_t i = 0; i < window->out_capacity; i++) {
+		double real, imag, mag, addition;
+		real = window->out_buffer[i][0];
+		imag = window->out_buffer[i][1];
+		mag = real * real + imag * imag;
```

Comment:
Should there be a `sqrt` here, or is it deliberately omitted for performance?

If we want performance and don't care too much about slight errors, there is the [alpha max plus beta min](https://en.wikipedia.org/wiki/Alpha_max_plus_beta_min_algorithm) hypotenuse approximation.

Reply:
AIUI (from https://www.osti.gov/servlets/purl/5688766 page 16, the most intelligible paper on welch's method I could find), we find the absolute magnitude of the fourier transform's result and immediately square it, so I skipped taking the square root and immediately squaring it.

Reply:
Oh, I didn't spot a subsequent squaring when I was reviewing the code!